src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,248 @@
-""" This is a dummy example """
+import pandas as pd
+import numpy as np
+import requests
+import matplotlib.pyplot as plt
+import time
+import logging
+from requests.models import Response
+
+
+API_URL = "https://climate-api.open-meteo.com/v1/climate?"
+COORDINATES = {
+    "Madrid": {"latitude": 40.416775, "longitude": -3.703790},
+    "London": {"latitude": 51.507351, "longitude": -0.127758},
+    "Rio": {"latitude": -22.906847, "longitude": -43.172896},
+}
+VARIABLES = ["temperature_2m_mean", "precipitation_sum", "soil_moisture_0_to_10cm_mean"]
+MODELS = [
+    "CMCC_CM2_VHR4",
+    "FGOALS_f3_H",
+    "HiRAM_SIT_HR",
+    "MRI_AGCM3_2_S",
+    "EC_Earth3P_HR",
+    "MPI_ESM1_2_XR",
+    "NICAM16_8S",
+]
+MAX_CALL_ATTEMPTS = 50
+
+# Logging configuration
+logging.basicConfig(
+    level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s"
+)
+
+
+def validate_response(response: Response) -> bool:
+    if not isinstance(response, dict):
+        return False
+
+    # Check if the required keys are in the response
+    required_keys = [
+        "latitude",
+        "longitude",
+        "generationtime_ms",
+        "timezone",
+        "timezone_abbreviation",
+        "daily",
+        "daily_units",
+    ]
+    if not all(key in response for key in required_keys):
+        return False
+
+    # Check if 'daily' is a dictionary
+    if not isinstance(response["daily"], dict):
+        return False
+
+    # Check if 'daily_units' is a dictionary
+    if not isinstance(response["daily_units"], dict):
+        return False
+
+    return True
+
+
+def call_api(params: dict):
+    """
+    Makes an API call with given parameters, handling rate limits and other errors.
+    :param params: Dictionary with the parameters to be passed to the API call.
+    :return: The API response if successful, None otherwise.
+    """
+    call_attempts = 0
+    backoff_time = 60
+    while call_attempts < MAX_CALL_ATTEMPTS:
+        try:
+            response = requests.get(API_URL, params=params)
+            if response.status_code == 200:
+                if validate_response(response.json()):
+                    return response
+            elif response.status_code == 429:
+                # Use the Retry-After header to determine the cool-off time
+                cool_off_time = response.headers.get("Retry-After", backoff_time)
+                logging.info(
+                    f"Rate limit exceeded. Waiting for {cool_off_time} seconds."
+                )
+                logging.info(response.text)
+                time.sleep(cool_off_time)
+                backoff_time *= 2
+            else:
+                raise Exception(
+                    f"API request failed with status code {response.status_code}"
+                )
+        except Exception as e:
```

Comment:
Not good practice to cover ALL exceptions. Be more granular. In this [response](https://stackoverflow.com/questions/16511337/correct-way-to-try-except-using-python-requests-module) of Stackoverflow you can find a good summary of the exceptions to cover.